### PR TITLE
[40171] Add VarHandleOp to set of kernels not to cache

### PR DESCRIPTION
**Pull request by frreiss**
Following the discussion with @jaingaurav on pull request #42337 , I have reimplemented the patch for the fourth memory leak from issue #40171 . This new version of the patch filters out `VarHandleOp` operations from the set of operations whose kernels are cached inside `EagerExecutor`. Excluding `VarHandleOp` from caching prevents `saved_model.load()` from leaking kernels when loading variable values. This change reduces the amount of memory leakage on my test script (which repeatedly loads a toy Keras model) substantially:

cache_vs_no_cache

Not caching these kernels should have minimal negative impact on performance as far as I'm aware.  `VarHandleOp::Compute()` is a relatively lightweight operation that usually runs right before a more expensive operation.

### CODE REVIEW

**Review by jaingaurav — APPROVED** [review]

**Review by Ayush-Parhi — APPROVED** [review]

**Comment by allenlavoie** [comment]
Sorry for the delay. @jaingaurav and I discussed this a bit; the fix for tf.Variable was to always use a fixed shared_name to indicate an anonymous resource. The reason something is still leaking is this line: https://github.com/tensorflow/tensorflow/blob/6d57022510f5820cfc3c5b86ba52fe623351bc63/tensorflow/python/ops/resource_variable_ops.py#L1957

Which should be like this line instead: https://github.com/tensorflow/tensorflow/blob/6d57022510f5820cfc3c5b86ba52fe623351bc63/tensorflow/python/ops/resource_variable_ops.py#L1706

The advantage is that we can keep the VarHandleOp kernels cached. I don't have a great sense of how much difference that'd make initializing models with lots of small variables, but it seems like it's worth having for the same reason we cached kernels in the first place.

Does that sound reasonable to you? I have a fix pending just changing that line (+ adding a heap check unit test for UninitializedVariable).

**Comment by frreiss** [comment]
@allenlavoie I suspect that change would produce some unintended sharing of buffers if the user were to load the same model twice in a single session. But maybe I'm reading the resource manager code incorrectly.

**Comment by allenlavoie** [comment]
Right, that's the "to indicate an anonymous resource" bit. The branching is here: https://github.com/tensorflow/tensorflow/blob/2ab4f849127adff4d9445200cf7c518a0620f437/tensorflow/core/kernels/resource_variable_ops.cc#L226

That logic we've had for a while. The change for UninitializedVariable (which SavedModel uses, vs. ResourceVariable which does not leak for this reason) is https://github.com/tensorflow/tensorflow/commit/84ff8673dca1a9b764069e189415bf76b34623d9

At least for CPU we're running a heapchecker and it's passing after that change; previously it did identify the kernel cache as a leak. I haven't tried it on a larger SavedModel load, but presumably leaks wouldn't be VarHandleOp kernel cache related anymore.

**Comment by gbaned** [comment]
@frreiss  Can you please check @allenlavoie's comments and keep us posted ? Thanks!

**Comment by frreiss** [comment]
Ah, I see now. Thanks for clarifying, @allenlavoie !

**Comment by gbaned** [comment]
@frreiss  Any update on this PR? Please. Thanks!

**Comment by allenlavoie** [comment]
I think we can close this for now.
